contracts/src/Teleporter/upgrades/TeleporterRegistry.sol

```diff
@@ -0,0 +1,74 @@
+// (c) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+// SPDX-License-Identifier: Ecosystem
+
+pragma solidity 0.8.18;
+
+import "../../WarpProtocolRegistry.sol";
+import "../ITeleporterMessenger.sol";
+
+/**
+ * @dev TeleporterRegistry contract is a {WarpProtocolRegistry} and provides an upgrade
+ * mechanism for teleporter messenger contracts.
+ */
+contract TeleporterRegistry is WarpProtocolRegistry {
+    mapping(address => uint256) internal _addressToVersion;
+
+    error DuplicateProtocolAddress();
+
+    constructor(
+        uint256[] memory initialVersions,
+        address[] memory initialProtocolAddresses
+    ) WarpProtocolRegistry(initialVersions, initialProtocolAddresses) {}
+
+    /**
+     * @dev Gets the {ITeleporterMessenger} contract of the given `version`.
+     */
+    function getTeleporter(
+        uint256 version
+    ) external view returns (ITeleporterMessenger) {
+        return ITeleporterMessenger(_getVersionToAddress(version));
+    }
+
+    /**
+     * @dev Gets the latest {ITeleporterMessenger} contract.
+     */
+    function getLatestTeleporter()
+        external
+        view
+        returns (ITeleporterMessenger)
+    {
+        return ITeleporterMessenger(_getVersionToAddress(_latestVersion));
+    }
+
+    /**
+     * @dev Gets the version of the given `protocolAddress`.
+     * If `protocolAddress` is not a valid protocol address, returns 0.
+     */
+    function getAddressToVersion(
+        address protocolAddress
+    ) external view returns (uint256) {
+        return _addressToVersion[protocolAddress];
+    }
+
+    /**
+     * @dev See {WarpProtocolRegistry-_addToRegistry}
+     *
+     * Adds the new protocol version address to the `_addressToVersion` mapping
+     * so that the registry can be queried for the version of a given protocol address.
+     * Requirements:
+     *
+     * - `protocolAddress` must not have been previously registered.
```

Comment:
Curious what the motivation for this requirement is? IIUC, I think this explicitly prohibits a "rollback" to a previously deployed version, which could be useful in a hypothetical situation where a new version is deployed to introduce a new feature, but contains a regression.

Reply:
I think we should also explicitly document whether or not changing the address of an existing version is supported. In the current implementation, it looks like this is supported, but we may want to consider disallowing it. Doing so would introduce a similar discrepancy issue between chains where one chain thinks the latest version is X and the other thinks it is Y.

Reply:
hmm I was originally thinking that this would just prevent a subnet from incorrectly adding a protocol address that was previously registered to avoid duplicate, and since they can just set their Teleporter version to an existing version in the case of a regression. But thinking again, I'll likely remove this duplicate protocol address check, so that in the case of a regression we can add a previous version address as the new version, and all dapps will continue with old address when using `getLatestTelepoter`

Reply:
I'm assuming you mean change the address of an existing version in the registry? This isn't allowed, when we call `_addToRegistry` the version being registered has to be `_latestVersion+1`, meaning new versions have always be incrementing by 1, and can't specify old version numbers.

Reply:
removed the check for duplicate protocol address and the error

Reply:
> I'm assuming you mean change the address of an existing version in the registry? This isn't allowed, when we call _addToRegistry the version being registered has to be _latestVersion+1, meaning new versions have always be incrementing by 1, and can't specify old version numbers.

Gotcha, sounds good. Yeah I missed that previously. Would you mind adding that requirement as a comment here, explaining it's enforced by the `WarpProtocolRegistry` implementation?

Reply:
added